assets/js/base/utils/render-frontend.tsx

```diff
@@ -0,0 +1,256 @@
+	render(
+		<BlockErrorBoundary { ...errorBoundaryProps }>
+			<Suspense fallback={ <div className="wc-block-placeholder" /> }>
+				{ Block && <Block { ...props } attributes={ attributes } /> }
+			</Suspense>
+		</BlockErrorBoundary>,
+		container
+	);
+};
+
+interface RenderBlockInContainersParams<
+	TProps extends Record< string, unknown >,
+	TAttributes extends Record< string, unknown >
+> {
+	// React component to use as a replacement.
+	Block: BlockType< TProps, TAttributes > | null;
+	// Containers to replace with the Block component.
+	containers: NodeListOf< Element >;
+	// Function to generate the props object for the block.
+	getProps?: GetPropsFn< TProps, TAttributes >;
+	// Function to generate the props object for the error boundary.
+	getErrorBoundaryProps?: (
+		el: HTMLElement,
+		i: number
+	) => Record< string, unknown >;
+}
+
+/**
+ * Renders a block component in each `containers` node.
+ */
+const renderBlockInContainers = <
+	TProps extends Record< string, unknown >,
+	TAttributes extends Record< string, unknown >
+>( {
+	Block,
+	containers,
+	getProps = () => ( {} as BlockProps< TProps, TAttributes > ),
+	getErrorBoundaryProps = () => ( {} ),
+}: RenderBlockInContainersParams< TProps, TAttributes > ): void => {
+	if ( containers.length === 0 ) {
+		return;
+	}
+
+	// Use Array.forEach for IE11 compatibility.
+	Array.prototype.forEach.call( containers, ( el, i ) => {
+		const props = getProps( el, i );
+
+		const errorBoundaryProps = getErrorBoundaryProps( el, i );
+		const attributes = {
+			...el.dataset,
+			...( props.attributes || {} ),
+		};
+		el.classList.remove( 'is-loading' );
+
+		renderBlock( {
+			Block,
+			container: el,
+			props,
+			attributes,
+			errorBoundaryProps,
+		} );
+	} );
+};
+
+// Given an element and a list of wrappers, check if the element is inside at
+// least one of the wrappers.
+const isElementInsideWrappers = (
+	el: Element,
+	wrappers: NodeListOf< Element >
+): boolean => {
+	return Array.prototype.some.call(
+		wrappers,
+		( wrapper ) => wrapper.contains( el ) && ! wrapper.isSameNode( el )
+	);
+};
+
+interface RenderBlockOutsideWrappersParams<
+	TProps extends Record< string, unknown >,
+	TAttributes extends Record< string, unknown >
+> extends RenderFrontendParams< TProps, TAttributes > {
+	// All elements matched by the selector which are inside the wrapper will be ignored.
+	wrappers?: NodeListOf< Element >;
+}
+
+/**
+ * Renders the block frontend in the elements matched by the selector which are
+ * outside the wrapper elements.
+ */
+const renderBlockOutsideWrappers = <
+	TProps extends Record< string, unknown >,
+	TAttributes extends Record< string, unknown >
+>( {
+	Block,
+	getProps,
+	getErrorBoundaryProps,
+	selector,
+	wrappers,
+}: RenderBlockOutsideWrappersParams< TProps, TAttributes > ): void => {
+	const containers = document.body.querySelectorAll( selector );
+	// Filter out blocks inside the wrappers.
+	if ( wrappers && wrappers.length > 0 ) {
```

Comment:
Can we simplify to `wrappers?.length > 0`?

Reply:
Typescript doesn't like it :'( 
![image](https://user-images.githubusercontent.com/4463174/144615918-a9024061-2f0a-4bd8-9ebe-aec97eaf9966.png)